integr8ly/integreatly-operator
# MGDAPI-3891 Added metric and alert to track failed APIManagementTenant CRs

# Issue link
JIRA: https://issues.redhat.com/browse/MGDAPI-3891

# What
We need an alert that fires whenever a APIManagementTenant CR fails to reconcile. This PR creates a new metric `num_failed_tenants` and a new alert `ApiManagementTenantCRFailed`. The alert is only created for multitenant installations of RHOAM.

# Verification steps
1. [Provision](https://qaprodauth.cloud.redhat.com/openshift/create) a non-CCS cluster
2. Checkout this PR and run `INSTALLATION_TYPE=multitenant-managed-api make cluster/prepare/local` on the command line once the cluster is ready
3. Create a CatalogSource either using `oc` or the GUI:
```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: rhmi-operators
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/ckyrillo/managed-api-service-index:1.22.0
```
4. Once the CatalogSource's Status is `READY`, install the RHOAM operator through the OperatorHub (OLM installation). Make sure to change the Installed Namespace to `sandbox-rhoam-operator`
5. Once the operator is installed, create a RHMI CR by running `INSTALLATION_TYPE=multitenant-managed-api USE_CLUSTER_STORAGE=true make deploy/integreatly-rhmi-cr.yml` on the command line
6. Wait until the the RHMI CR's Status.Stage is `Complete` and then setup the testing IDP by running `DEDICATED_ADMIN_PASSWORD=Password1 PASSWORD=Password1 ./scripts/setup-sso-idp.sh` on the command line
7. Once the IDP is deployed on the cluster, login `test-user01` through the `testing-idp` so a user is created
8. Next create the following Project either using `oc` or the GUI: `test-user01-dev`
9. Login into the `sandbox-rhoam-observability` instance of Prometheus and confirm that the alert `ApiManagementTenantCRFailed` **is not** firing
10. Now create an APIManagementTenant CR in the `test-user01-dev` namespace:
```yaml
apiVersion: integreatly.org/v1alpha1
kind: APIManagementTenant
metadata:
  name: example
  namespace: test-user01-dev
spec: {}
```
11. Wait until the CR has reconciled and confirm that the alert is still not firing.
12. Once the above CR has reconciled and the alert is confirmed to not be firing, create another CR:
```yaml
apiVersion: integreatly.org/v1alpha1
kind: APIManagementTenant
metadata:
  name: this-should-fail
  namespace: test-user01-dev
spec: {}
```
13. The second CR's `ProvisioningStatus` field should be `won't provision`
14. Lastly, go back to Prometheus and confirm that the alert **is** firing (it might take a minute)
